iib/web/api_v1.py

```diff
@@ -170,22 +171,44 @@ def get_build_logs(request_id):
     :raise Gone: if the logs for the build request have been removed due to expiration
     """
     request_log_dir = flask.current_app.config['IIB_REQUEST_LOGS_DIR']
-    if not request_log_dir:
+    s3_bucket_name = flask.current_app.config['IIB_AWS_S3_BUCKET_NAME']
+    if not s3_bucket_name and not request_log_dir:
         raise NotFound()
 
     request = Request.query.get_or_404(request_id)
-    log_file_path = os.path.join(request_log_dir, f'{request_id}.log')
-    if not os.path.exists(log_file_path):
+
+    finalized = request.state.state_name in RequestStateMapping.get_final_states()
+    if not finalized:
+        raise ValidationError(
+            f'The request {request_id} is not complete yet.'
+            ' logs will be available once the request is complete.'
+        )
+
+    # If S3 bucket is configured, fetch the log file from the S3 bucket.
+    # Esle, check if logs are stored on the system itself and return them.
+    # Otherwise, raise an IIBError.
+    if s3_bucket_name:
+        log_file = get_file_from_s3_bucket(f'request_logs/{request_id}.log', s3_bucket_name)
```

Comment:
Typo on `Else` hehe

Besides that, here's the confirmation that we're storing the S3 object using `key_prefix` and `file_prefix` but retrieving them as a single concatenated parameter. Not sure if this is the best approach, IMO is inconsistent.